File: Developpement/SwalkitAndroid/app/src/main/java/fr/insarennes/ih2a/swalkitandroid/SwalkitSignal.kt
package fr.insarennes.ih2a.swalkitandroid

import fr.insarennes.ih2a.swalkitandroid.data.DataSignal

data class SwalkitSignal(var intensity:Int, var pulse:Int, var distance:Int) {
    companion object {
        fun intensityString(value:Int) = "${value}%"
        fun pulseString(value:Int) = if (value > 0) "${value}ms" else "OFF"
        fun distanceString(value:Int) = "${value}cm"
    }

    fun intensityString() = Companion.intensityString(intensity)
    fun pulseString() = Companion.pulseString(pulse)
    fun distanceString() = Companion.distanceString(distance)

    override fun toString(): String {
        return "${intensityString()}, ${pulseString()}, ${distanceString()}"
    }

    fun toDataSignal() : DataSignal = DataSignal(intensity, pulse, distance)
    fun fromDataSignal(dataSignal:DataSignal) {
        intensity = dataSignal.intensity
        pulse = dataSignal.pulse
        distance = dataSignal.distance
    }
}

File: Developpement/Swalkit_firmware_M5atomS3/src/SwalkitSignal.cpp
#include "SwalkitSignal.h"
#include <sstream>


void SwalkitSignal::fromBytes(const uint8_t *data) {
    intensity = static_cast<int>(data[0]);
    pulse = static_cast<int>(data[1]) * 10;
    distance = static_cast<int>(data[2]);
}
    
void SwalkitSignal::toBytes(uint8_t *data) {
    data[0] = static_cast<uint8_t>(intensity);
    data[1] = static_cast<uint8_t>(pulse / 10);
    data[2] = static_cast<uint8_t>(distance);
}

std::string SwalkitSignal::toString() {
    std::ostringstream oss;
    oss << intensity << ", " << pulse << ", " << distance;
    return oss.str();
}
